### PR TITLE
Issue #244  Initial draft of performance testing page including summarized case study of DSPT use of APDEX

### PR DESCRIPTION
**Pull request by chrisclarknhsnet**
As per discussions with Andy

### CODE REVIEW

File: practices/performancetesting.md [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+<div style="padding-right:30px;padding-bottom:30px;">
+Previously we had defined a list of user scenarios for the typical actions undertaken on the system which we named after Mr Men characters.  We also defined for every 100 users how many (i.e. what percentage) would be likely to be performing a given Mr Man scenario
+</div>
+We used these scenarios to define our thread groups within JMeter and decided we would run our performance tests for 250 users at a time which would constitute a heavy load for the system.
+
+### A case study (what does good look like?)
+
+We decided that we wanted to know if the performance for a particular scenario had degraded by more than 5% compared to previous average performance.  If it did we wanted to fail the pipeline so we could investigate any new pieces of code further.
+
+### A case study (approach)
```
Comment (andyblundell) [inline practices/performancetesting.md:78]: Please could you add a brief section on the *impact* in this case study, e.g. perhaps a process that used to take X hours of person time and therefore was only run every X weeks now requires no human intervention, and is run every day/week/release-candidate/whatever?

**Review by andyblundell — CHANGES_REQUESTED** [review]
This is great - will re-review tomorrow, in the meantime have suggested one extra section. Is there any code from the case study that would be worth sharing? Maybe the Jenkins config snippet? Obviously your jmeter config is bespoke.

File: practices/performancetesting.md [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+Although you can apply APDEX figures to JMeter it only calculates them per endpoint whereas we wanted to aggregate our APDEX figures at the Mr Man scenario level.
+
+We therefore wrote a Python program which would take the raw JMeter results file (a sample of shown below) and using regular expressions to group results by thread names matching a Mr Man scenario, would calculate the aggregate APDEX score per scenario.
```
Comment (stefaniuk) [inline practices/performancetesting.md:82]: Would you be happy to include your Python code and configuration that could be helpful to others as part of the case study?

**Review by stefaniuk — DISMISSED** [review]
Would it be ok to change the location of the new images and the naming convention for the files, please? E.g. `practices/images/dspt-case-study-aggregated-apdex-scores.png`

**Comment by stefaniuk** [comment]
This case study reads really well. Nice work!

File: practices/performancetesting.md [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+* Would our users become unhappy when there is an increased load?
+
+## A case study
+
+For the Data Security Protection Toolkit (DSPT) we decided to use APDEX so that, prior to a fortnightly release, we could answer the question:
+
+> Has this release made the performance of the system worse?
+
+### A case study (know your audience)
+<div style="float:right">
```
Comment (andyblundell) [inline practices/performancetesting.md:67]: I'm not sure this is rendering how it was supposed to...?

Comment (chrisclarknhsnet) [reply practices/performancetesting.md:67]: Amended to just use simple mark up, that'll teach me,

File: practices/performancetesting.md (hunk 2 above)
Comment (chrisclarknhsnet) [reply practices/performancetesting.md:86]: Yes, although I probably want to give it a quick sanity check first as was hacking it around a bit a few weeks ago.
Will look to add this early next week.

Comment (andyblundell) [reply practices/performancetesting.md]: Converted to https://github.com/NHSDigital/software-engineering-quality-framework/issues/246 for the code/config to be added (hopefully) later - will resolve this comment in the meantime

@stefaniuk FYI

File: practices/performance-testing.md [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+* It potentially hides an individual page or end point whose performance has degraded due to the aggregation/smoothing effect of APDEX.
+* Recognition that we probably want in the future to also apply an absolute target e.g. APDEX of >= 0.9
+
+### A case study - architecture
+
+The following diagram summarises the approach taken to using APDEX by DSPT:
+
+![Alt](./case-studies/performance-dspt/images/architecture.png "DSPT Performance test architecture")
+
+### A case study - impact
```
Comment (andyblundell) [inline practices/performance-testing.md:116]: @chrisclarknhsnet are you happy with this final section please? I think it needed a summary / impact to close, and I've tried to paraphrase content from the case study - hopefully I've not misrepresented you!

Comment (chrisclarknhsnet) [reply practices/performance-testing.md:116]: Yep that's fine Andy

**Review by andyblundell — APPROVED** [review]

**Pull request merged**